### PR TITLE
Extend declgen with support for logicals, save and initial values

### PR DESCRIPTION
**Pull request by arporter**
This PR is for work done in #172 and is needed by #170.

### CODE REVIEW

File: src/psyclone/f2pygen.py [hunk 1]
```diff
@@ -37,8 +37,7 @@
     provide routines which can be used to generate fortran code. This library
     includes pytest tests. '''
 
-import fparser
-import fparser.one
+from __future__ import absolute_import
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: I'm not sure why this is required. It makes no difference with our without it as far as I can tell.

File: src/psyclone/f2pygen.py [hunk 2]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+            reader.set_format(fort_fmt)
+            myline = reader.next()
+            self._decl = fparser1.typedecl_statements.Logical(parent.root,
+                                                              myline)
         else:
-            raise RuntimeError(
-                "f2pygen:DeclGen:init: Only integer and real are currently"
-                " supported and you specified '{0}'".format(datatype))
-        # make a copy of entity_decls as we may modify it
+            # Defensive programming in case SUPPORTED_TYPES is added to
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: Good. We should always include such checks in if statements as it makes adding new functionality so much easier.

File: src/psyclone/f2pygen.py [hunk 3]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+            raise NotImplementedError(
+                "Internal error: type '{0}' is in DeclGen.SUPPORTED_TYPES "
+                "but not handled by constructor.".format(dtype))
+
+        # Make a copy of entity_decls as we may modify it
         local_entity_decls = entity_decls[:]
-        self._decl.entity_decls = local_entity_decls
+        if initial_values:
+            # Create a list of 2-tuples
+            value_pairs = zip(local_entity_decls, initial_values)
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: nice use of zip, although I don't really like the name of the function as it implies compression to me.

File: src/psyclone/f2pygen.py [hunk 4]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+            # Construct an assignment from each tuple
+            self._decl.entity_decls = ["=".join(_) for _ in value_pairs]
+        else:
+            self._decl.entity_decls = local_entity_decls
+
         my_attrspec = []
         if intent != "":
             my_attrspec.append("intent({0})".format(intent))
-        if pointer is not False:
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: What terrible programming by the previous coder!

File: src/psyclone/f2pygen.py [hunk 5]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+        if pointer:
             my_attrspec.append("pointer")
-        if allocatable is not False:
+        if allocatable:
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: And again, shameful.

File: src/psyclone/f2pygen.py [hunk 6]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+    def _check_initial_values(dtype, values):
+        '''
+        Check that the supplied values are consistent with the requested
+        data type.
+        :param str dtype: Fortran intrinsic type
+        :param list values: list of values as strings
+        :raises RuntimeError: if the supplied values are not consistent
+                              with the specified data type
+        '''
+        from fparser.two.pattern_tools import abs_name, \
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: This is something a "full" parser would be able to check.

File: src/psyclone/f2pygen.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/f2pygen.py]: I've no idea why I added it either - I can't see any difference with and without. I shall remove it.

File: src/psyclone/f2pygen.py [hunk 7]
```diff
@@ -766,62 +765,148 @@ def __init__(self, parent, content):
...
+    :param str kind: the KIND attribute to use for this declaration
+    :param str dimension: the DIMENSION specifier (i.e. the xx in
+                          DIMENSION(xx))
+    :param bool allocatable: whether this declaration is for an
+                             ALLOCATABLE quantity
+    :param bool save: whether this declaration should have the SAVE attribute
+    :param initial_values: Initial value to give each variable.
+    :type initial_values: list of str with same no. of elements as entity_decls
+    :raises RuntimeError: if no variable names are specified
+    :raises RuntimeError: if datatype is not one of "integer" or "real"
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/f2pygen.py]: Oops. It should.

File: src/psyclone/f2pygen.py (hunk 6 above)
Comment (arporter) [reply src/psyclone/f2pygen.py]: Good plan, I shall do that.

File: src/psyclone/tests/f2pygen_test.py [hunk 8]
```diff
@@ -889,7 +916,105 @@ def test_adduse_default_funcnames():
...
-def test_declgen_wrong_type():
+def test_decl_logical():
+    ''' Check that we can create a declaration for a logical variable '''
+    module = ModuleGen(name="testmodule")
+    sub = SubroutineGen(module, name="testsubroutine")
+    module.add(sub)
+    sub.add(DeclGen(sub, datatype="logical", entity_decls=["first_time"]))
+    gen = str(sub.root).lower()
+    assert "logical first_time" in gen
+    # Add a second logical variable
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/f2pygen_test.py]: Done.

File: src/psyclone/tests/f2pygen_test.py [hunk 9]
```diff
@@ -889,7 +916,105 @@ def test_adduse_default_funcnames():
...
+def test_decl_save():
+    ''' Check that we can declare variables with the save attribute '''
+    module = ModuleGen(name="testmodule")
+    sub = SubroutineGen(module, name="testsubroutine")
+    module.add(sub)
+    for idx, dtype in enumerate(DeclGen.SUPPORTED_TYPES):
+        sub.add(DeclGen(sub, datatype=dtype, save=True,
+                        entity_decls=["var"+str(idx)]))
+    gen = str(sub.root).lower()
+    assert "logical, save :: var" in gen
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/f2pygen_test.py]: Good point - done.

**Comment by arporter** [comment]
Ready for 2nd review...

**Pull request merged**